xtm/xtm_scsp_queue.h

```diff
@@ -0,0 +1,184 @@
+ *    following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above
+ *    copyright notice, this list of conditions and the following
+ *    disclaimer in the documentation and/or other materials
+ *    provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY <COPYRIGHT HOLDER> ``AS IS'' AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+ * TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+ * <COPYRIGHT HOLDER> OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
+ * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
+ * BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+ * LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+ * THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+
+/**
+ * Lock free single consumer, single producer queue,
+ * based on ring buffer.
+ * Template parameter can be of any type for which the
+ * copying operation is carried out quickly and correctly.
+ */
+template<class T>
+struct xtm_scsp_queue {
+	/** Next position to be written */
+	unsigned write;
+	/** Next position to be read */
+	unsigned read;
+	/** Circular buffer length */
+	unsigned len;
+	/** Buffer contains objects */
+	T buffer[];
+	/**
+	 * Init scsp queue struct.
+	 * @param[in] size - size of queue ring buffer, must be power of two,
+	 *                   for fast calculation of read/write position, using
+	 *                   '& (size - 1' operation.
+	 * @retval 0 if success, otherwise return -1.
+	 */
+	inline int
+	xtm_scsp_queue_init(unsigned size) {
+		/* Ensure size is power of 2 */
+		if (size & (size - 1))
+			return -1;
+
+		write = 0;
+		read = 0;
+		len = size;
+		return 0;
+	}
+	/**
+	 * Add num elements into queue
+	 * @param[in] data - array of elements to add
+	 * @param[in] num - count of elements to add
+	 * @retval number of elements actually written.
+	 */
+	inline unsigned
+	xtm_scsp_queue_put(T *data, unsigned num) {
+		unsigned i;
+		unsigned queue_write = write;
+		unsigned new_write = queue_write;
+		unsigned len_minus_1 = len - 1;
+		unsigned queue_read = __atomic_load_n(&read, __ATOMIC_ACQUIRE);
+
+		for (i = 0; i < num; i++) {
+			new_write = (new_write + 1) & len_minus_1;
+			if (new_write == queue_read)
+				break;
+			buffer[queue_write] = data[i];
+			queue_write = new_write;
+		}
+		__atomic_store_n(&write, queue_write, __ATOMIC_RELEASE);
+		return i;
+	}
+	/**
+	 * Get up to num elements from the queue.
+	 * @param[out] data - array to save elements
+	 * @param[in] num - count of elements to get
+	 * @retval number of elements actually read.
+	 */
+	inline unsigned
+	xtm_scsp_queue_get(T *data, unsigned num) {
```

Comment:
The function is never used (or did I miss its usage somewhere?).

Reply:
Yes It's remained from the original idea with a void * queue. I didn't delete it because i think may be this queue implementation will be useful in other places? If not i delete this function!

Reply:
Up to you. But if you leave it, it would be good to cover it with tests.

Reply:
I removed this feature until a suitable user case appears